### ISSUE TITLE
[strict-boolean-expressions] False negative (missing error)

### ISSUE THREAD

**Issue by OliverJAsh** [typescript-eslint/typescript-eslint#2666]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

Note: this is a duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/1118, which was marked as fix but it looks like it is not fixed.

Code: https://github.com/OliverJAsh/typescript-eslint-strict-boolean-expressions-bug

`.eslintrc.js`:
```js
const config = {
  parser: "@typescript-eslint/parser",
  parserOptions: {
    project: ["./tsconfig.json"],
  },
  plugins: ["@typescript-eslint"],
  rules: {
    "@typescript-eslint/strict-boolean-expressions": 2,
  },
};

module.exports = config;
```

`main.ts`:
```ts
declare const myString: string;
declare const myBoolean: boolean;

const expression = myBoolean && myString;
if (expression) {
} // Expect: error. Actual: error. ✅

if (myBoolean && myString) {
} // Expect: error. Actual: no error. ❌
```

`tsconfig.json`:
```json
{
    "compilerOptions": {
        "strict": true
    }
}
```

**Expected Result**

See comments in code above.

**Actual Result**

See comments in code above.

**Additional Info**

**Versions**

`package.json`:
```json
{
  "dependencies": {
    "@typescript-eslint/eslint-plugin": "^4.4.0",
    "@typescript-eslint/parser": "^4.4.0",
    "eslint": "^7.11.0",
    "typescript": "^4.0.3"
  }
}
```

**Comment by bradzacher** [comment]
So this is actually working as expected - though it might not be entirely clear why.

Note that the default config has `allowString: true`.

Starting with the second case:
```ts
if (myBoolean && myString) {
}
```
The algorithm is:
1) check `if` test
  a) it's a logical expression, so check every part of it 
2) check left of logical
  a) it's an identifier, so just check the type of it
  b) its type is `boolean` - always considered "safe".
3) is this an `if` test? Yes so check right of logical
  a) it's an identifier, so just check the type of it
  b) its type is `string`
  c) is `allowString` true? Yes so this is "safe".
4) done.

Because your logical expression is inlined in the `if` test, we can interrogate its parts and clearly determine that every part of the logical is correctly handled.

For the second case:
```ts
const expression = myBoolean && myString;
if (expression) {
}
```
The algorithm is:
1) check `if` test
  a) it's an identifier, so check the type of it
  b) its type is `boolean | string`
  c) there's no setting which allows this type, so error

---

To summarise:
- the reason that we error for the first case is because `boolean | string` is not considered to be strict, because there are multiple truthy and multiple falsey cases.
- the reason that we don't error for the second case is because we can interrogate the inline logical expression and determine that whilst its return type is `boolean | string`, each case is strictly handled according to the config.

**Comment by OliverJAsh** [comment]
Thanks for providing such a detailed answer.

> Note that the default config has `allowString: true`.

Ahhh so I completely missed this. I thought it defaulted to `false`.

Out of interest, is it not possible to interrogate the various types inside the condition regardless of whether the condition is inlined or not? This way the behaviour would be consistent.

**Comment by bradzacher** [comment]
> Out of interest, is it not possible to interrogate the various types inside the condition regardless of whether the condition is inlined or not?

We could interrogate the union easily, but in this instance the reason we report here is because there are multiple falsey (`false` and `''`) and multiple truthy (`true` and not empty string) states, so it's considered an unstrict expression (i.e. did you mean to just check for the true case? or did you mean to just check for the non-empty string case? or did you mean to check for both?)

OTOH it's because we can deeply interrogate the logical expression itself (_not_ its types) that we can tell that both cases are handled independently and strictly (i.e. we can tell you specifically and intentionally checked for the true case and the non-empty string case).

You get the same brand of verification from TS - it does not do control-flow analysis on related variables, even though (as a developer) it's clear how it's runtime-safe:
```ts
declare const x: string | null;
declare const y: string | null;

if (x && y) {
    // x & y are non-null
}

const logical = x && y;
if (logical) {
    // x & y are still null
}

```

**Comment by OliverJAsh** [comment]
Ahh interesting. Thank you!
